### PR TITLE
Tell IPython the correct GUI event loop to use for all backends.

### LINKED ISSUE #10763: Cairo in 2.2.0 not working for new backends
### Bug report

**The new Cairo backends do not produce plots**

I tried the new Cairo backends, Qt4Cairo, Qt5Cairo, WXCairo, and TkCairo, but neither of them produced plots for me.  (Fedora 27, compiled own python 3.6.4 from source, all extra packages installed from pip3).

**Code for reproduction**

matplotlibrc:
```
backend      : TkCairo
# all other lines commented out
```
code:

```python
from matplotlib.pylab import *
plot([0,1])
show()
```

**Actual outcome**

Qt4Cairo, Qt5Cairo - frame but no plot, frame not updated
WXCairo, TkCairo - no frame

Just to be sure, GTK3Cairo makes a frame with a plot, all working fine.
Same for Qt5Agg, WXAgg, TkAgg all work as well

Qt4Agg is broken ("ModuleNotFoundError: No module named 'PyQt4'", there seems to be no such module on pip)

So it seems neither Cairo, WX, Tk, QT5 are broken by themselves, but the integration does not work smoothly, at least in my setup.  I have no clue how to debug this.

**Expected outcome**

A plot.

**Matplotlib version**
  * Operating system: Fedora 27
  * Matplotlib version: 2.2.0
  * Matplotlib backend (see above)
  * Python version: 3.6.4 
  * IPython: 6.2.1
  
python compiled from source, packages (current) from pip3

### PR DESCRIPTION
**Pull request by anntzer**
IPython currently uses a hard-coded table
(IPython.core.pylabtools.backend2gui) to know which event loop to use
for which backend.  This approach fails for both the new builtin
cairo-based backends (which do not appear in the table), and for
third-party backends (e.g. mplcairo).

mplcairo has used some custom code to patch that table for a while;
reuse it to more generally use the new "required_interactive_framework"
attribute.

Note that this PR suggests that there should be a better way to go back
from the canvas class to the backend module (rather than looking for
`sys.modules[cls.__module__].required_interactive_framework`, which is a
bit hacky and could in theory fail if the canvas class is actually
defined in another submodule).

(Edits:
Actually the "hack" above "sort-of" fails for Qt4 in the sense that `backend_qt4agg.FigureCanvas` is actually the *same* class as `backend_qt5agg.FigureCanvas`, so trying to reach the required_interactive_framework that way will point to "qt5".  But fortunately ipython maps both "qt4" and "qt5" to "qt", so we're fine...

Perhaps we can just mandate that the `__module__` attribute of FigureCanvas subclasses must point to their actual backend module name.  It's not as if requiring that the backend module name be stored in *another* attribute really helps.)

Fixes #10763.
attn @Carreau 

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/backend_bases.py [hunk 1]
```diff
@@ -1629,6 +1632,35 @@ def __init__(self, figure):
         self.toolbar = None  # NavigationToolbar2 will set me
         self._is_idle_drawing = False
 
+    @classmethod
+    @functools.lru_cache(1)
```
Comment (tacaswell) [inline lib/matplotlib/backend_bases.py]: Won't this cache be shared across all sub-classes?

**Comment by tacaswell** [comment]
This is probably the right way to do this, definitely need to negotiate with the IPython folks to make this a bit more of a documented interface.

File: lib/matplotlib/backend_bases.py (hunk 1 above)
Comment (anntzer) [reply lib/matplotlib/backend_bases.py]: Sorry, indeed it should be `lru_cache()` (an unbounded cache).

The lru_cache(1) comes from mplcairo which did this just once with the list of hardcoded mappings just for mplcairo (as it predates the required_interactive_framework attribute).

**Comment by Carreau** [comment]
I'm happy to get any patch in IPython that can help you. I suppose we should probably find a nice way to register new backends with IPython (entrypoints ?). I'm also limited in bandwidth since I'm not 100% jupyter-funded, so I'm happy to work on giving access to the repository/PYPI/.... etc if I become a blocking factor.

**Comment by anntzer** [comment]
> I suppose we should probably find a nice way to register new backends with IPython (entrypoints ?).

I don't think that really helps as Matplotlib's current policy is that *any* module can be a backend, as long as it defines the correct functions/classes/attributes.

I am reasonably happy with the solution in this PR (i.e., let the backend register itself to IPython's table when a canvas is about to be created (which is more or less the latest the registration can be done)).  Alternatively, IPython could itself have the code that tries to read `required_interactive_framework`, but that seems worse: 1. it would need to update itself if the backend is switched, and 2. it would neither work with old versions of Matplotlib that didn't have that attribute, nor with old versions of IPython which, well, didn't do that; whereas the solution here "works" for all versions of IPython that have the "more-or-less public" :-) API that this touches.

**Comment by anntzer** [comment]
rebased

**Comment by tacaswell** [comment]
I'm in favor of merging this as long as @Carreau does not tell is it is too hacky.

**Comment by jklymak** [comment]
@Carreau did you have a minute to look at this?

**Comment by jklymak** [comment]
@Carreau did not tell us it was too hacky.  If it is, I imagine we can back it out....

**Pull request merged**